### PR TITLE
Replace Datadog/zstd with Klauspost/compress

### PR DESCRIPTION
**Pull request by jarifibrahim**
Fixes https://github.com/dgraph-io/badger/issues/1162
PR https://github.com/dgraph-io/badger/pull/1176 was the first attempt at replacing the library but we decided to not switch the library because of performance issues.

https://github.com/klauspost/compress has become much faster since the last time we benchmarked it in https://github.com/dgraph-io/badger/pull/1176. The new benchmarks in https://github.com/dgraph-io/badger/issues/1162#issuecomment-646958797 look very promising and we should get rid of the CGO dependency. 

---
This change is [Reviewable](https://reviewable.io/reviews/dgraph-io/badger/1383)

### CODE REVIEW

**Comment by klauspost** [comment]
> The block size of 4KB was converted to a size of 5554 bytes.

Let me double check. That shouldn't be possible.

File: y/zstd.go [hunk 1]
```diff
@@ -19,18 +17,38 @@
...
+	var err error
+	zstdEncOnce.Do(func() {
+		zstdEnc, err = zstd.NewWriter(
+			nil, zstd.WithZeroFrames(true),
+			zstd.WithEncoderLevel(zstd.EncoderLevelFromZstd(level)),
+			zstd.WithEncoderCRC(false))
+
+		AssertTrue(err == nil)
+	})
+	return zstdEnc.EncodeAll(src, dst), nil
```
Comment (klauspost) [inline y/zstd.go:53]:
```suggestion
	return zstdEnc.EncodeAll(src, dst[:0]), nil
```

This is likely the problem. Output is appended to make it easier to use in serializers, etc.

File: y/zstd.go [hunk 2]
```diff
@@ -19,18 +17,38 @@
...
 
 // ZSTDDecompress decompresses a block using ZSTD algorithm.
 func ZSTDDecompress(dst, src []byte) ([]byte, error) {
-	return zstd.Decompress(dst, src)
+	var err error
+	zstdDecOnce.Do(func() {
+		zstdDec, err = zstd.NewReader(nil)
+		AssertTrue(err == nil)
+	})
+	return zstdDec.DecodeAll(src, dst)
```
Comment (klauspost) [inline y/zstd.go:39]:
```suggestion
	return zstdDec.DecodeAll(src, dst[:0])
```
If you don't want it appended.

**Comment by ivanjaros** [comment]
Any progress with this?

**Comment by klauspost** [comment]
@jarifibrahim Ping. You might want to upgrade to `v1.11.3`, there are some minor tweaks and fixes. I don't see them invalidating the tests, but might as well get them in.

**Comment by emolitor** [comment]
Has this PR been abandoned? Slightly confused by the state of the ticket as reading through the comments it appears all the issues which were raised have been addressed. Are there other issues which are not documented or is this ticket just waiting for time by the author to test and commit?

**Comment by NamanJain8** [comment]
@jarifibrahim closing this because the issue is addressed by #1706

**Comment by jarifibrahim** [comment]
We have merged https://github.com/dgraph-io/badger/pull/1706 which adds the pure go-based ZSTD to badger. This will be released as part of the badger release in July.

@mholt @emolitor @ivanjaros @klauspost Thank you all for your feedback and helping improve badger :heart: . Apologies for the delay. This PR was lost in all the other work we're doing on badger/dgraph.

**Comment by jarifibrahim** [comment]
@francislavoie Okay. I think we can set pure go as the default one. I think @klauspost already mentioned that the klauspost/compress should be able to decompress data written by datadog/zstd. So we shouldn't have a problem.

**Comment by klauspost** [comment]
Yes and vice versa if that should become relevant.

**Pull request closed without merging**